queue/queue.go

```diff
@@ -0,0 +1,303 @@
+	Kind        queueType
+	Opts
+}
+
+func (cfg Cfg) toString() string {
+	params := []string{fmt.Sprintf("temporary = %v, if_not_exists = %v", cfg.Temporary, cfg.IfNotExists)}
+
+	if cfg.Ttl.Seconds() != 0 {
+		params = append(params, fmt.Sprintf("ttl = %f", cfg.Ttl.Seconds()))
+	}
+
+	if cfg.Ttr.Seconds() != 0 {
+		params = append(params, fmt.Sprintf("ttr = %f", cfg.Ttr.Seconds()))
+	}
+
+	if cfg.Delay.Seconds() != 0 {
+		params = append(params, fmt.Sprintf("delay = %f", cfg.Delay.Seconds()))
+	}
+
+	return "{" + strings.Join(params, ",") + "}"
+}
+
+func (cfg Cfg) getType() string {
+	kind := string(cfg.Kind)
+	if kind == "" {
+		kind = string(FIFO)
+	}
+
+	return kind
+}
+
+type Opts struct {
+	Pri   int           // task priorities
+	Ttl   time.Duration // task time to live
+	Ttr   time.Duration // task time to execute
+	Delay time.Duration // delayed execution
+}
+
+func (opts Opts) toMap() map[string]interface{} {
+	ret := make(map[string]interface{})
+
+	if opts.Ttl.Seconds() != 0 {
+		ret["ttl"] = opts.Ttl.Seconds()
+	}
+
+	if opts.Ttr.Seconds() != 0 {
+		ret["ttr"] = opts.Ttr.Seconds()
+	}
+
+	if opts.Delay.Seconds() != 0 {
+		ret["delay"] = opts.Delay.Seconds()
+	}
+
+	if opts.Pri != 0 {
+		ret["pri"] = opts.Pri
+	}
+
+	return ret
+}
+
+// Connect creates and configures new Connection
+func Connect(addr string, opts tarantool.Opts) (*Connection, error) {
+	conn, err := tarantool.Connect(addr, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Connection{conn, opts}, nil
+}
+
+// New creates a new queue with config and return Queue.
+func (conn *Connection) NewQueue(name string, cfg Cfg) (Queue, error) {
+	var q queue
+	cmd := fmt.Sprintf("queue.create_tube('%s', '%s', %s)", name, cfg.getType(), cfg.toString())
+	_, err := conn.c.Eval(cmd, []interface{}{})
+	if err == nil {
+		q = queue{
+			name,
+			conn,
+			makeCmdMap(name),
+		}
+	}
+	return q, err
+}
+
+// GetQueue returns an existing queue by name.
+func (conn *Connection) GetQueue(name string) (Queue, error) {
+	var q queue
+	cmd := fmt.Sprintf("return queue.tube.%s ~= null", name)
+	resp, err := conn.c.Eval(cmd, []interface{}{})
+	if err != nil {
+		return q, err
+	}
+
+	exist := len(resp.Data) != 0 && resp.Data[0].(bool)
+
+	if exist {
+		q = queue{
+			name,
+			conn,
+			makeCmdMap(name),
+		}
+	} else {
+		err = fmt.Errorf("Tube %s does not exist", name)
+	}
+
+	return q, err
+}
+
+// Close closes Connection.
+func (conn *Connection) Close() error {
+	return conn.c.Close()
+}
+
+// Put data to queue. Returns task.
+func (q queue) Put(data interface{}) (*Task, error) {
+	return q.put(data)
+}
+
+// Put data with options (ttl/ttr/pri/delay) to queue. Returns task.
+func (q queue) PutWithOpts(data interface{}, cfg Opts) (*Task, error) {
+	return q.put(data, cfg.toMap())
+}
+
+func (q queue) put(p ...interface{}) (*Task, error) {
+	var (
+		params []interface{}
+		task   *Task
+	)
+	params = append(params, p...)
+	resp, err := q.conn.c.Call(q.cmd["put"], params)
```

Comment:
I don't demand this change. If you like it then apply. If you don't, I wouldn't mind.

````go
type cnt int                                                                                                                                                                                                                              
func (c *cnt) inc(add bool) {                                                                                                                                                                                                             
    if add {                                                                                                                                                                                                                              
        *c++                                                                                                                                                                                                                              
    }                                                                                                                                                                                                                                     
}                                                                                                                                                                                                                                         
                                                                                                                                                                                                                                          
func (o Opts) EncodeMsgpack(e msgpack.Encoder) error {                                                                                                                                                                                    
    cnt := cnt(0)                                                                                                                                                                                                                         
    cnt.add(opts.Ttl != 0)                                                                                                                                                                                                                
    cnt.add(opts.Ttr != 0)                                                                                                                                                                                                                
    cnt.add(opts.Delay != 0)                                                                                                                                                                                                              
    cnt.add(opts.Pri != 0)                                                                                                                                                                                                                
    if cnt == 0 {                                                                                                                                                                                                                         
        return e.EncodeNil()                                                                                                                                                                                                              
    }                                                                                                                                                                                                                                     
    e.EncodeMapLen(int(cnt))                                                                                                                                                                                                       
    if opts.Ttl != 0 {                                                                                                                                                                                                                    
        e.EncodeString("ttl")                                                                                                                                                                                                             
        e.EncodeFloat64(opts.Ttl.Seconds())                                                                                                                                                                                               
    }                                                                                                                                                                                                                                     
    if opts.Ttr != 0 {                                                                                                                                                                                                                    
        e.EncodeString("ttr")                                                                                                                                                                                                             
        e.EncodeFloat64(opts.Ttr.Seconds())                                                                                                                                                                                               
    }                                                                                                                                                                                                                                     
    if opts.Delay != 0 {                                                                                                                                                                                                                  
        e.EncodeString("delay")                                                                                                                                                                                                           
        e.EncodeFloat64(opts.Delay.Seconds())                                                                                                                                                                                             
    }                                                                                                                                                                                                                                     
    if opts.Pri != 0 {                                                                                                                                                                                                                    
        e.EncodeString("pri")                                                                                                                                                                                                             
        e.EncodeInt(opts.Pri)                                                                                                                                                                                                             
    }                                                                                                                                                                                                                                     
    return nil                                                                                                                                                                                                                            
}                                                                                                                                                                                                                                         
                                                                                                                                                                                                                                          
func (q queue) Put(data interface{}) (*Task, error) {                                                                                                                                                                                     
    return q.PutWithOpts(data, Opts{})                                                                                                                                                                                                    
}                                                                                                                                                                                                                                         
                                                                                                                                                                                                                                          
// Put data with options (ttl/ttr/pri/delay) to queue. Returns task.                                                                                                                                                                      
func (q queue) PutWithOpts(data interface{}, cfg Opts) (*Task, error) {                                                                                                                                                                   
    var task *Task                                                                                                                                                                                                                        
    params := []interface{}{ data, nil }                                                                                                                                                                                                  
    if cfg != Opts{} {                                                                                                                                                                                                                    
        params[1] = cfg                                                                                                                                                                                                                   
    }                                                                                                                                                                                                                                     
    resp, err := q.conn.c.Call(q.cmd["put"], params)
````